.github/workflows/test.yaml

```
@@ -25,11 +30,66 @@ jobs:
             go.sum
       # run tests on PR only if there are code changes and always on main
       - name: Run Go Tests
-        if: ${{ (steps.changed-files.outputs.any_changed == 'true'  && github.event_name == 'pull_request') || github.ref == 'refs/heads/main' }}
+        if: ${{ steps.changed-files.outputs.any_changed == 'true' && github.event_name == 'pull_request' || github.ref == 'refs/heads/main'}}
         run: |
           make test
+      # upload raw coverage report always if tests ran
       - name: Archive code coverage results
         uses: actions/upload-artifact@v4
+        if: ${{ steps.changed-files.outputs.any_changed == 'true' && github.event_name == 'pull_request' || github.ref == 'refs/heads/main'}}
         with:
           name: code-coverage
           path: cover.out
+     # upload html coverage report only on PR and if tests ran
+      - name: Upload html coverage report
+        uses: actions/upload-artifact@v4
+        if: ${{ steps.changed-files.outputs.any_changed == 'true' && github.event_name == 'pull_request'}}
+        id: html-artifact
+        with:
+          name: code-coverage-html
+          path: cover.html
+      # create comment with coverage html report url on PR and if tests ran
+      - uses: actions/github-script@v6
+        name: Create comment with html report
+        if: ${{ steps.changed-files.outputs.any_changed == 'true' && github.event_name == 'pull_request'}}
+        with:
+          script: |
+            const commentIdentifier = "### Coverage HTML report";
+
+            // Get all comments on the pull request
+            const { data: comments } = await github.rest.issues.listComments({
+              issue_number: context.issue.number,
+              owner: context.repo.owner,
+              repo: context.repo.repo,
+            });
+
+            // Find the comment that contains the identifier
+            const existingComment = comments.find(comment => comment.body.includes(commentIdentifier));
+
+            const newBody = `${commentIdentifier} \n ${{ steps.html-artifact.outputs.artifact-url }}`;
+
+            if (existingComment) {
+              // Update the existing comment
+              await github.rest.issues.updateComment({
+                comment_id: existingComment.id,
+                owner: context.repo.owner,
+                repo: context.repo.repo,
+                body: newBody,
+              });
+            } else {
+              // Create a new comment
+              await github.rest.issues.createComment({
+                issue_number: context.issue.number,
+                owner: context.repo.owner,
+                repo: context.repo.repo,
+                body: newBody,
+              });
+            }
+
+      # create comment with coverage changes digest on PR and if tests ran
+      - name: "Code coverage report"
+        uses: fgrosse/go-coverage-report@v1.0.2
```

Optional, there's newer version of the action:

```suggestion
        uses: fgrosse/go-coverage-report@v1.2.0
```